### PR TITLE
feat: add custom event `beforeUpdateEvent`

### PR DESCRIPTION
**Pull request by lja1018**
### Please check if the PR fulfills these requirements
- [x] It's submitted to right branch according to our branching model
- [x] It's right issue type on title
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] It does not introduce a breaking change or has description for the breaking change

### Description

- add custom event `beforeUpdateEvent`
- display form popup when edit button is clicked
- remove mergedRef hooks

- example for popup arrow in form popup

https://user-images.githubusercontent.com/28647773/149270753-f4cc8300-a0f7-4b52-bcc9-78fcf18cb523.mov

---
Thank you for your contribution to TOAST UI product. 🎉 😘 ✨

### CODE REVIEW

File: apps/calendar/src/components/popup/eventFormPopup.tsx [hunk 1]
```diff
@@ -12,59 +12,119 @@ import { EventStateSelector } from '@src/components/popup/eventStateSelector';
...
 
+function getChanges(event: EventModel, eventModelData: EventModelData) {
+  return Object.entries(eventModelData).reduce((changes, [key, value]) => {
+    const eventModelDataKey = key as keyof EventModelData;
+
+    if (event[eventModelDataKey] instanceof TZDate) {
+      // NOTE: handle TZDate
+      if (compare(event[eventModelDataKey], value) !== 0) {
+        changes[eventModelDataKey] = value;
+      }
```
Comment (lja1018) [inline apps/calendar/src/components/popup/eventFormPopup.tsx:82]: 객체가 `TZDate`, `raw` 필드밖에 없어 `TZDate`만 예외처리함.

**Comment by adhrinae** [comment]
@lja1018 description에 올린 영상이 잘못된 것 같은데 확인 부탁해요.

**Comment by lja1018** [comment]
@adhrinae 
넵. 사실 화살표가 찍히는걸 찍은건데 애매하게 됬네요. 본문 내용에 추가하겠습니다.

File: apps/calendar/src/components/popup/eventDetailPopup.tsx [hunk 2]
```diff
@@ -34,23 +30,25 @@ function calculatePopupPosition(eventRect: Rect, layoutRect: Rect, popupRect: Re
   let top = eventRect.top + eventRect.height / 2 - popupRect.height / 2;
   let left = eventRect.left + eventRect.width;
 
-  if (top + popupRect.height > layoutRect.top + layoutRect.height) {
+  if (isTopOverLayoutContainer(top, layoutRect, popupRect)) {
```
Comment (adhrinae) [inline apps/calendar/src/components/popup/eventDetailPopup.tsx:33]: `is~OutOf~` 가 더 어울릴 것 같네요.

File: apps/calendar/src/components/popup/eventDetailPopup.tsx [hunk 3]
```diff
@@ -34,23 +30,25 @@ function calculatePopupPosition(eventRect: Rect, layoutRect: Rect, popupRect: Re
...
 
-  return [Math.max(0, top), Math.max(left, layoutRect.left)];
+  return [Math.max(top, layoutRect.top), Math.max(left, layoutRect.left)];
 }
 
 function calculatePopupArrowPosition(eventRect: Rect, layoutRect: Rect, popupRect: Rect) {
   const top = eventRect.top + eventRect.height / 2;
   const popupLeft = eventRect.left + eventRect.width;
 
   const isOverLayoutContainer = popupLeft + popupRect.width > layoutRect.left + layoutRect.width;
```
Comment (adhrinae) [inline apps/calendar/src/components/popup/eventDetailPopup.tsx:48]: `isOutOfLayoutContainer`

File: apps/calendar/src/components/popup/eventDetailPopup.tsx [hunk 4]
```diff
@@ -123,7 +127,19 @@ export const EventDetailPopup: FunctionComponent = () => {
...
+            event={event}
+            title={title}
+            location={location}
+            start={start}
+            end={end}
+            isAllday={isAllday}
+            isPrivate={isPrivate}
+            popupArrowPointPosition={popupArrowPointPosition}
+          />
+        )}
```
Comment (adhrinae) [inline apps/calendar/src/components/popup/eventDetailPopup.tsx:142]: 굳이 별도의 컴포넌트로 분리하지 않아도 되겠습니다.
이 컴포넌트 안에서 한번에 다 처리해보죠.

**Review by adhrinae — CHANGES_REQUESTED** [review]

**Pull request merged**